### ISSUE TITLE
[ServiceBus] Is there documentation on how to set up retry or timeout configuration?

### ISSUE THREAD

**Issue by yunbozhang-msft** [Azure/azure-sdk-for-go#22629]
### Service Package Request

Hi team, 

When I used the ServiceBus SDK, I found no documentation to guide me in setting the timeout and retry strategy. Are these strategies already built in?
For example, after I fail to send a message, will the SDK automatically help me resend it? Or I need to implement re-try by myself?

Thanks!

**Comment by richardpark-msft** [comment]
> Thanks for the feedback! We are routing this to the appropriate team for follow-up. cc @EldertGrootenboer @jfggdl.

This is purely an SDK issue, no need for @EldertGrootenboer, @jfggdl :)

**Comment by richardpark-msft** [comment]
@zhangyunbo1994, we've got an example which I think is what you're asking for:

https://github.com/Azure/azure-sdk-for-go/blob/afea09b57d9701e17e45e13fdeb444ccdda85138/sdk/messaging/azservicebus/example_client_test.go#L45

It's also in our pkg.go.dev docs but associated with NewClient, which might make it less discoverable: [link](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus#example-NewClient-ConfiguringRetries)

> For example, after I fail to send a message, will the SDK automatically help me resend it? Or I need to implement re-try by myself?

We use this retry policy for exactly the case you mentioned (and others). This also applies to just reconnecting, etc..

There are some errors that are not considered retryable (for instance, failed authentication) but other things like timeouts we will use the SDK configured timeouts that you set above.

**Comment by richardpark-msft** [comment]
@zhangyunbo1994 - btw, let me know how I can make that more discoverable, if possible!

**Comment by yunbozhang-msft** [comment]
Thanks for your answer!

**Comment by richardpark-msft** [comment]
Ah, https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-samples#go-samples isn't optimal. I talked to the docs person, we're going to change it to `https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/messaging/azservicebus#pkg-examples`, which includes the retry example I showed you earlier, and also provides a better list than what you'd get by just looking at the GitHub repo.

As to your second part, about a "Best Practices" guide, that's a great idea and one we are missing across all the repos. From a general perspective, the library is already built to support best practices, from a library perspective - it includes retries, exponential backoff and, automatically provides for reconnecting on connection disruption. 

There are some other cases that are worth calling out though - for instance, using the same `azservicebus.Client` allows you to manage the # of TCP connections used, and re-using an existing Receiver or Sender is more efficient than constantly creating and closing them. I'll create a separate issue so I can collate some more ideas.

For now though I think, given the example, and the doc update, that we're done with specific issue. I'll close this and open another to track the best practices bit.
